### PR TITLE
Add Shibboleth providers

### PR DESCRIPTION
**Pull request by OmeGak**

### CODE REVIEW

File: example/example.cfg.example [hunk 1]
```diff
@@ -27,6 +28,12 @@ MULTIAUTH_AUTH_PROVIDERS = {
     'local': {
         'type': 'example_local',
         'title': 'Local Accounts'
+    },
+    'my_shibboleth': {
+        'type': 'shibboleth',
+        'title': 'Shibboleth',
+        'callback_uri': '/shibboleth/sso',
```
Comment (ThiefMaster) [inline example/example.cfg.example]: I wouldn't put this in the example config. In most cases the generated name is fine and documenting what the URL will be and that it needs to be configured in the shibboleth section of the Apache config should be enough.

Another option would be removing the default value for it and raising an exception if it's not defined... Having to explicitly specify the URL in any case wouldn't be so bad...

File: example/example.cfg.example [hunk 2]
```diff
@@ -27,6 +28,12 @@ MULTIAUTH_AUTH_PROVIDERS = {
...
+        'attr_prefix': 'ADFS_'
```
Comment (ThiefMaster) [inline example/example.cfg.example]: I wouldn't put default values in here.

File: example/example.wsgi [hunk 3]
```diff
@@ -0,0 +1,150 @@
+# This file is part of Flask-MultiAuth.
```
Comment (ThiefMaster) [inline example/example.wsgi]: NO

File: flask_multiauth/providers/shibboleth.py [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+
+    This provider expects to find a Shibboleth module configured in
+    Apache server running the application.
+    """
+
+    def __init__(self, *args, **kwargs):
+        super(ShibbolethAuthProvider, self).__init__(*args, **kwargs)
+        self.settings.setdefault('attrs_prefix', 'ADFS_')
+        self.settings.setdefault('callback_uri', '/shibboleth/{}'.format(self.name))
+        self.authorized_endpoint = '_flaskmultiauth_shibboleth_' + self.name
```
Comment (ThiefMaster) [inline flask_multiauth/providers/shibboleth.py]: `shibboleth_endpoint` maybe? "authorized" was quite oauth specific

File: flask_multiauth/providers/shibboleth.py [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+from flask_multiauth.data import AuthInfo, IdentityInfo
+from flask_multiauth.exceptions import AuthenticationFailed, IdentityRetrievalFailed
+from flask_multiauth.identity import IdentityProvider
+from flask_multiauth.util import login_view
+
+
+class ShibbolethAuthProvider(AuthProvider):
+    """Provides authentication for Shibboleth.
+
+    This provider expects to find a Shibboleth module configured in
```
Comment (ThiefMaster) [inline flask_multiauth/providers/shibboleth.py]: This sentence is rather bad english

File: flask_multiauth/providers/shibboleth.py [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+class ShibbolethIdentityProvider(IdentityProvider):
+    """Provides identity information for Shibboleth
+
+    This provider expects to find a Shibboleth module configured in
+    Apache server running the application.
+    """
+
+    def __init__(self, *args, **kwargs):
+        super(ShibbolethIdentityProvider, self).__init__(*args, **kwargs)
+        self.settings.setdefault('identifier_field', 'ADFS_PERSONID')
```
Comment (ThiefMaster) [inline flask_multiauth/providers/shibboleth.py]: Should we really provide a default here. I think PERSONID is very CERN specific.

File: example/example.cfg.example (hunk 1 above)
Comment (OmeGak) [inline example/example.cfg.example]: It's something you have to configure manually in the Shibboleth config anyway, so I'm not against it.

File: flask_multiauth/providers/sqlalchemy.py [hunk 7]
```diff
@@ -4,7 +4,7 @@
 # Flask-MultiAuth is free software; you can redistribute it
 # and/or modify it under the terms of the Revised BSD License.
 
-from __future__ import unicode_literals
+from __future__ import unicode_literals, absolute_import
```
Comment (ThiefMaster) [inline flask_multiauth/providers/sqlalchemy.py]: That's already done in jacques' ldap branch

File: example/example.wsgi (hunk 3 above)
Comment (OmeGak) [inline example/example.wsgi]: :+1: I was just feeling dirty for committing this whole file that it's pretty much a copy of the other.

Comment (OmeGak) [inline example/example.wsgi]: :volcano: 
> Symbolic link not allowed or link target not accessible: /opt/indico/flask-multiauth/example/example.wsgi

File: flask_multiauth/providers/shibboleth.py (hunk 6 above)
Comment (OmeGak) [inline flask_multiauth/providers/shibboleth.py]: `ADFS_LOGIN` sounds more like a sensible default.

File: flask_multiauth/providers/sqlalchemy.py (hunk 7 above)
Comment (OmeGak) [inline flask_multiauth/providers/sqlalchemy.py]: Well, we need it anyway for this PR and it shouldn't cause any conflicts.

File: example/example.py [hunk 8]
```diff
@@ -64,7 +64,7 @@ def identity_handler(identity_info):
     if not identity:
         user = User.query.filter_by(email=identity_info.data['email']).first()
         if not user:
-            user = User(**identity_info.data)
+            user = User(**identity_info.data.to_dict())
```
Comment (ThiefMaster) [inline example/example.py]: Could be in a separate commit.. it's not related to shibboleth

**Comment by ThiefMaster** [comment]
Merged in 624c640
